### PR TITLE
use hardened xml parser defusedxml

### PR DESCRIPTION
**Pull request by JanZerebecki**
It took entirely too long, but currently released OS distributions now have a safe by default ElementTree, but I think this may be used by some people on older systems that are still vulnerable by shipping it via ssh to that old system from a new system with cockpit.

The xml here comes from something that is in a normal configuration not untrusted (a process running as a local user receives the xml from the systems dbus/systemd instance) and it doesn't cross a possible security boundary between cockpit-bridge and cockpit-ws, as this is probably only used in cockpit-bridge. On the other hand this is published on pypi.

I considered if this is not needed, overall if this change just works, it is probably still a good idea.

### CODE REVIEW

**Comment by allisonkarlitskaya** [comment]
Hi!

Thanks for the contribution.  I'm currently leaning against the proposal for a few reasons, most of which probably aren't surprising:
 - systemd_ctypes is meant to be vendored by other projects (ie: Cockpit) and adding a non-stdlib dependency to it would complicate that relationship
 - third party dependencies consumed via pypi come with their own security concerns (as you noted)
 - (also as you note) this problem is not a problem with recent Python versions
 - (also as you note) xml parsing is *probably* not a security boundary for most uses of this code.  If we're connecting to the user bus, it's to programs running as the same user, and if we're connecting to a well-known name on the the system bus, it's a program that the sysadmin has configured.  I might imagine a situation where we get tricked into communicating via unique name with an untrustworthy process running as a non-trusted user on the system bus, or with (for example) a sandboxed flatpak on the user bus, but given the other considerations above, I don't consider this to be a very large point.

We might also choose to deal with this situation by using one of the other xml parsers in the stdlib if they have better security properties, but if this is really just about older versions, I think I'd prefer to avoid doing that sort of work as well.

I'd like to confer with other members of the Cockpit team before a final decision, which will probably happen next week.

Thanks!

**Comment by allisonkarlitskaya** [comment]
From a practical standpoint, we care about RHEL 8 (and its many derivatives):

From RHEL 8.4 and 8.10:

```python
>>> pyexpat.EXPAT_VERSION
'expat_2.2.5'
```

which suggests that we have a vulnerable version of expat.  And indeed, I tried subjecting it to a billion laughs via the `ElementTree` API in Python and it chews CPU until the OOM killer comes for it...

The same test performed on my Fedora 40 system, even via Python 3.6, produces this message fairly quickly:

```
xml.etree.ElementTree.ParseError: limit on input amplification factor (from DTD and entities) breached: line 15, column 6
```

So: it's not about the Python version so much as the expat version, and RHEL 8 is definitely vulnerable.

**Comment by JanZerebecki** [comment]
We could vendor defusedxml:
It is using the hidden pure python parser in stdlib. Other than that is there any other one in stdlib that is not using expat? defusedxml docs imply no. This is using less than 200 lines from defusedxml, so it is relatively easy to vendor. Instead of trying to reduce the size or reimplement defusedxml it is probably best to just copy 3 files from it.

Which makes me ask: Why not rebuild python against a new version of expat, or fix python to default to the pure python implemenation if an expat with a fixed API is not available? Going around patching users of an API that just for defaults seems very non-productive.

**Comment by allisonkarlitskaya** [comment]
> Which makes me ask: Why not rebuild python against a new version of expat, or fix python to default to the pure python implemenation if an expat with a fixed API is not available? Going around patching users of an API that just ask for defaults seems very non-productive.

I agree.  Even with the old expat version number on RHEL 8 I was surprised to see that it's still vulnerable... I'd have figured that it got patched by now....

**Review by martinpitt — COMMENTED** [review]
TBH I'm not convinced about the "security" aspect of this. Users can always pwn themselves (user bus), and if someone can pwn the system bus, they effectively already won. Also, the kind of XML that makes ET go :fire: isn't the kind you usually find in D-Bus introspections.

I am rather strongly against an external dependency. I could accept 
```py
try:
    import defusedxml.ElementTree as ET
except ImportError:
    import xml.etree.ElementTree as ET
```

but it increases (ideally, doubles) the testing space and introduces a potential incompatibility/bugs into the API.

So in summary, I'm rather against this TBH.

**Comment by allisonkarlitskaya** [comment]
I think we're gonna turn this one down, sorry.

Please let us know if you find any other issues you think are worth addressing!

**Comment by JanZerebecki** [comment]
For future reference, there are some cpython issues still open related to this:
- https://github.com/python/cpython/issues/61441
- https://github.com/python/cpython/issues/68426 
- https://github.com/python/cpython/issues/63682

Seems like that is lacking people, nobody working on this for years now.
